File: frontend/src/api/demo/modelSignalsDrift.ts
import type {
  ModelSignalDrift,
  ModelSignalDriftMetric,
  ModelSignalMetricSet,
  ModelSignalRates
} from '../types'

type DriftSeverity = 'critical' | 'warning'

interface DriftSignal {
  severity: DriftSeverity
  key: string
  label: string
  direction: string
  reason: string
  currentValue: number
  baselineValue: number
}

interface ThresholdSignalInput extends Omit<DriftSignal, 'severity'> {
  change: number
  criticalAt: number
  warningAt: number
  enabled?: boolean
}

interface ConditionalSignalInput extends DriftSignal {
  condition: boolean
}

function relativeIncrease(current: number, baseline: number): number {
  if (baseline <= 0) return current > 0 ? 1 : 0
  return (current - baseline) / baseline
}

function relativeDecrease(current: number, baseline: number): number {
  if (baseline <= 0) return 0
  return (baseline - current) / baseline
}

export function reasoningOverhead(metric: Pick<ModelSignalRates, 'reasoningOverheadRate'>): number {
  return metric.reasoningOverheadRate
}

export function modelSignalDriftFor(current: ModelSignalMetricSet, baseline: ModelSignalMetricSet): ModelSignalDrift {
  const signals = driftSignalsFor(current, baseline)
  const severity = signals.reduce((highest, signal) =>
    severityRank(signal.severity) > severityRank(highest) ? signal.severity : highest, 'healthy'
  )

  return {
    severity,
    confidence: current.sessionCount < 2 || current.modelCalls < 3 ? 'low' : 'high',
    sampleNote: current.sessionCount < 2 || current.modelCalls < 3 ? 'Low sample' : undefined,
    reasons: [...new Set(signals.map((signal) => signal.reason))],
    metrics: signals.map(signalMetric)
  }
}

function driftSignalsFor(current: ModelSignalMetricSet, baseline: ModelSignalMetricSet): DriftSignal[] {
  const reasoningOverheadDelta = reasoningOverhead(current) - reasoningOverhead(baseline)
  const degradationRiskDelta = current.degradationRiskScore - baseline.degradationRiskScore
  return [
    latencyDriftSignal(current, baseline),
    throughputDriftSignal(current, baseline),
    outputThroughputDriftSignal(current, baseline),
    toolFailureDriftSignal(current, baseline),
    cacheMissDriftSignal(current, baseline),
    reasoningOverheadDriftSignal(current, baseline, reasoningOverheadDelta),
    degradationRiskDriftSignal(current, baseline, degradationRiskDelta)
  ].filter((signal): signal is DriftSignal => signal !== undefined)
}

function latencyDriftSignal(current: ModelSignalMetricSet, baseline: ModelSignalMetricSet): DriftSignal | undefined {
  return thresholdSignal({
    change: relativeIncrease(current.modelLatencyMsPer1kOutputTokens, baseline.modelLatencyMsPer1kOutputTokens),
    criticalAt: 0.55,
    warningAt: 0.22,
    key: 'modelLatencyMsPer1kOutputTokens',
    label: 'model latency per 1k output tokens',
    direction: 'higher_worse',
    reason: 'Latency rose vs baseline',
    currentValue: current.modelLatencyMsPer1kOutputTokens,
    baselineValue: baseline.modelLatencyMsPer1kOutputTokens
  })
}

function throughputDriftSignal(current: ModelSignalMetricSet, baseline: ModelSignalMetricSet): DriftSignal | undefined {
  return thresholdSignal({
    change: relativeDecrease(current.modelThroughputTokensPerSecond, baseline.modelThroughputTokensPerSecond),
    criticalAt: 0.42,
    warningAt: 0.2,
    key: 'modelThroughputTokensPerSecond',
    label: 'model throughput',
    direction: 'lower_worse',
    reason: 'Throughput fell vs baseline',
    currentValue: current.modelThroughputTokensPerSecond,
    baselineValue: baseline.modelThroughputTokensPerSecond
  })
}

function outputThroughputDriftSignal(current: ModelSignalMetricSet, baseline: ModelSignalMetricSet): DriftSignal | undefined {
  return thresholdSignal({
    change: relativeDecrease(current.modelThroughputOutputTokensPerSecond, baseline.modelThroughputOutputTokensPerSecond),
    criticalAt: 0.5,
    warningAt: 0.24,
    key: 'modelThroughputOutputTokensPerSecond',
    label: 'model output throughput',
    direction: 'lower_worse',
    reason: 'Output throughput fell',
    currentValue: current.modelThroughputOutputTokensPerSecond,
    baselineValue: baseline.modelThroughputOutputTokensPerSecond
  })
}

function toolFailureDriftSignal(current: ModelSignalMetricSet, baseline: ModelSignalMetricSet): DriftSignal | undefined {
  return conditionalSignal({
    condition: current.failedToolCalls > baseline.failedToolCalls && current.toolFailureRate >= 0.08,
    severity: current.toolFailureRate >= 0.2 ? 'critical' : 'warning',
    key: 'toolFailureRate',
    label: 'tool failure rate',
    direction: 'higher_downstream_symptom',
    reason: 'Tool failures above baseline',
    currentValue: current.toolFailureRate,
    baselineValue: baseline.toolFailureRate
  })
}

function cacheMissDriftSignal(current: ModelSignalMetricSet, baseline: ModelSignalMetricSet): DriftSignal | undefined {
  return thresholdSignal({
    change: current.cacheMissRate - baseline.cacheMissRate,
    criticalAt: Number.POSITIVE_INFINITY,
    warningAt: 0.12,
    key: 'cacheMissRate',
    label: 'cache miss rate',
    direction: 'higher_symptom',
    reason: 'Cache misses above baseline',
    currentValue: current.cacheMissRate,
    baselineValue: baseline.cacheMissRate
  })
}

function reasoningOverheadDriftSignal(
  current: ModelSignalMetricSet,
  baseline: ModelSignalMetricSet,
  change: number
): DriftSignal | undefined {
  return thresholdSignal({
    change,
    criticalAt: Number.POSITIVE_INFINITY,
    warningAt: 0.12,
    key: 'reasoningOverheadRate',
    label: 'reasoning overhead',
    direction: 'cost_shape_review',
    reason: 'Reasoning overhead rose',
    currentValue: reasoningOverhead(current),
    baselineValue: reasoningOverhead(baseline)
  })
}

function degradationRiskDriftSignal(
  current: ModelSignalMetricSet,
  baseline: ModelSignalMetricSet,
  change: number
): DriftSignal | undefined {
  return thresholdSignal({
    change,
    criticalAt: 0.3,
    warningAt: 0.15,
    enabled: current.degradationRiskScore >= 0.3,
    key: 'degradationRiskScore',
    label: 'model quality risk score',
    direction: 'higher_worse',
    reason: 'Model quality risk rose',
    currentValue: current.degradationRiskScore,
    baselineValue: baseline.degradationRiskScore
  })
}

function thresholdSignal(input: ThresholdSignalInput): DriftSignal | undefined {
  if (input.enabled === false || input.change < input.warningAt) return undefined
  return {
    severity: input.change >= input.criticalAt ? 'critical' : 'warning',
    key: input.key,
    label: input.label,
    direction: input.direction,
    reason: input.reason,
    currentValue: input.currentValue,
    baselineValue: input.baselineValue
  }
}

function conditionalSignal(input: ConditionalSignalInput): DriftSignal | undefined {
  if (!input.condition) return undefined
  return {
    severity: input.severity,
    key: input.key,
    label: input.label,
    direction: input.direction,
    reason: input.reason,
    currentValue: input.currentValue,
    baselineValue: input.baselineValue
  }
}

function signalMetric(signal: DriftSignal): ModelSignalDriftMetric {
  return {
    key: signal.key,
    label: signal.label,
    direction: signal.direction,
    severity: signal.severity,
    current: signal.currentValue,
    baseline: signal.baselineValue,
    delta: signal.currentValue - signal.baselineValue,
    deltaPct: signal.baselineValue > 0 ? (signal.currentValue - signal.baselineValue) / signal.baselineValue : 0
  }
}

export function severityRank(value?: string): number {
  const normalized = (value || '').toLowerCase()
  if (normalized === 'critical' || normalized === 'high') return 3
  if (normalized === 'warning' || normalized === 'medium') return 2
  if (normalized === 'watch' || normalized === 'low') return 1
  return 0
}

File: internal/query/model_signals_drift.go
package query

import (
	"sort"
	"strings"

	"github.com/LyleMi/AgentMeter/internal/model"
)

type modelSignalDriftRuleKind int

const (
	modelSignalDriftRelativeIncrease modelSignalDriftRuleKind = iota
	modelSignalDriftRelativeDecrease
	modelSignalDriftAbsoluteIncrease
)

type modelSignalDriftField int

const (
	modelSignalDriftP90Latency modelSignalDriftField = iota
	modelSignalDriftLatency
	modelSignalDriftP10Throughput
	modelSignalDriftOutputThroughput
	modelSignalDriftToolFailureRate
	modelSignalDriftFailurePressure
	modelSignalDriftCacheMissRate
	modelSignalDriftAvgModelCalls
	modelSignalDriftOutputExpansion
	modelSignalDriftReasoningOverhead
	modelSignalDriftDegradationRisk
)

type modelSignalDriftRule struct {
	kind              modelSignalDriftRuleKind
	field             modelSignalDriftField
	key               string
	label             string
	direction         string
	reason            string
	warningThreshold  float64
	criticalThreshold float64
	minimumThreshold  float64
}

var modelSignalDriftRules = []modelSignalDriftRule{
	{
		kind:              modelSignalDriftRelativeIncrease,
		field:             modelSignalDriftP90Latency,
		key:               "p90ModelLatencyMsPer1kOutputTokens",
		label:             "p90 model latency per 1k output tokens",
		direction:         "higher_worse",
		reason:            "model latency increased",
		warningThreshold:  0.5,
		criticalThreshold: 1.0,
		minimumThreshold:  250,
	},
	{
		kind:              modelSignalDriftRelativeIncrease,
		field:             modelSignalDriftLatency,
		key:               "modelLatencyMsPer1kOutputTokens",
		label:             "model latency per 1k output tokens",
		direction:         "higher_worse",
		reason:            "model latency increased",
		warningThreshold:  0.5,
		criticalThreshold: 1.0,
		minimumThreshold:  250,
	},
	{
		kind:              modelSignalDriftRelativeDecrease,
		field:             modelSignalDriftP10Throughput,
		key:               "p10ModelThroughputTokensPerSecond",
		label:             "p10 model throughput",
		direction:         "lower_worse",
		reason:            "output throughput dropped",
		warningThreshold:  0.25,
		criticalThreshold: 0.5,
		minimumThreshold:  25,
	},
	{
		kind:              modelSignalDriftRelativeDecrease,
		field:             modelSignalDriftOutputThroughput,
		key:               "modelThroughputOutputTokensPerSecond",
		label:             "model output throughput",
		direction:         "lower_worse",
		reason:            "output throughput dropped",
		warningThreshold:  0.25,
		criticalThreshold: 0.5,
		minimumThreshold:  25,
	},
	{
		kind:              modelSignalDriftAbsoluteIncrease,
		field:             modelSignalDriftToolFailureRate,
		key:               "toolFailureRate",
		label:             "tool failure rate",
		direction:         "higher_downstream_symptom",
		reason:            "tool failure rate increased",
		warningThreshold:  0.10,
		criticalThreshold: 0.25,
		minimumThreshold:  0.10,
	},
	{
		kind:              modelSignalDriftAbsoluteIncrease,
		field:             modelSignalDriftFailurePressure,
		key:               "failurePressure",
		label:             "failure pressure",
		direction:         "higher_worse",
		reason:            "failure pressure increased",
		warningThreshold:  0.10,
		criticalThreshold: 0.25,
		minimumThreshold:  0.10,
	},
	{
		kind:              modelSignalDriftAbsoluteIncrease,
		field:             modelSignalDriftCacheMissRate,
		key:               "cacheMissRate",
		label:             "cache miss rate",
		direction:         "higher_symptom",
		reason:            "cache miss rate increased",
		warningThreshold:  0.20,
		criticalThreshold: 0.40,
		minimumThreshold:  0.50,
	},
	{
		kind:              modelSignalDriftRelativeIncrease,
		field:             modelSignalDriftAvgModelCalls,
		key:               "avgModelCallsPerSession",
		label:             "model calls per session",
		direction:         "higher_retry_loop_symptom",
		reason:            "model calls per session increased",
		warningThreshold:  0.5,
		criticalThreshold: 1.0,
		minimumThreshold:  0.5,
	},
	{
		kind:              modelSignalDriftRelativeIncrease,
		field:             modelSignalDriftOutputExpansion,
		key:               "outputExpansionRate",
		label:             "output expansion rate",
		direction:         "behavior_higher",
		reason:            "output expansion increased",
		warningThreshold:  1.0,
		criticalThreshold: 2.0,
		minimumThreshold:  1.0,
	},
	{
		kind:              modelSignalDriftAbsoluteIncrease,
		field:             modelSignalDriftReasoningOverhead,
		key:               "reasoningOverheadRate",
		label:             "reasoning overhead rate",
		direction:         "cost_shape_review",
		reason:            "reasoning overhead increased",
		warningThreshold:  0.50,
		criticalThreshold: 1.00,
		minimumThreshold:  0.50,
	},
	{
		kind:              modelSignalDriftAbsoluteIncrease,
		field:             modelSignalDriftDegradationRisk,
		key:               "degradationRiskScore",
		label:             "model quality risk score",
		direction:         "higher_worse",
		reason:            "model quality risk increased",
		warningThreshold:  0.15,
		criticalThreshold: 0.30,
		minimumThreshold:  0.30,
	},
}

func compareModelSignalDrift(current, baseline model.ModelSignalsMetricSet) model.ModelSignalsDrift {
	drift := model.ModelSignalsDrift{
		Severity:   modelSignalSeverityHealthy,
		Confidence: modelSignalConfidenceHigh,
		Reasons:    []string{},
		Metrics:    []model.ModelSignalsDriftMetric{},
	}
	if note := modelSignalSampleNote(current, baseline); note != "" {
		drift.Severity = modelSignalSeverityUnknown
		drift.Confidence = modelSignalConfidenceLow
		drift.SampleNote = note
		drift.Reasons = append(drift.Reasons, note)
		return drift
	}

	for _, rule := range modelSignalDriftRules {
		rule.addTo(&drift, current, baseline)
	}

	return drift
}

func (rule modelSignalDriftRule) addTo(drift *model.ModelSignalsDrift, current, baseline model.ModelSignalsMetricSet) {
	currentValue := rule.field.value(current)
	baselineValue := rule.field.value(baseline)
	metric, ok := rule.metric(currentValue, baselineValue)
	if !ok {
		return
	}
	appendModelSignalDriftMetric(drift, metric, rule.reason)
}

func (rule modelSignalDriftRule) metric(current, baseline float64) (model.ModelSignalsDriftMetric, bool) {
	delta := current - baseline
	deltaPct := safeDeltaPct(current, baseline)
	severity, ok := rule.severity(current, baseline, delta, deltaPct)
	if !ok {
		return model.ModelSignalsDriftMetric{}, false
	}
	return model.ModelSignalsDriftMetric{
		Key:       rule.key,
		Label:     rule.label,
		Direction: rule.direction,
		Severity:  severity,
		Current:   current,
		Baseline:  baseline,
		Delta:     delta,
		DeltaPct:  deltaPct,
	}, true
}

func (rule modelSignalDriftRule) severity(current, baseline, delta, deltaPct float64) (string, bool) {
	switch rule.kind {
	case modelSignalDriftRelativeIncrease:
		return rule.relativeIncreaseSeverity(current, baseline, delta, deltaPct)
	case modelSignalDriftRelativeDecrease:
		return rule.relativeDecreaseSeverity(current, baseline, delta, deltaPct)
	case modelSignalDriftAbsoluteIncrease:
		return rule.absoluteIncreaseSeverity(current, baseline, delta)
	default:
		return "", false
	}
}

func (rule modelSignalDriftRule) relativeIncreaseSeverity(current, baseline, delta, deltaPct float64) (string, bool) {
	if baseline <= 0 || current <= baseline || delta < rule.minimumThreshold {
		return "", false
	}
	if deltaPct >= rule.criticalThreshold {
		return modelSignalSeverityCritical, true
	}
	if deltaPct >= rule.warningThreshold {
		return modelSignalSeverityWarning, true
	}
	return "", false
}

func (rule modelSignalDriftRule) relativeDecreaseSeverity(current, baseline, delta, deltaPct float64) (string, bool) {
	if baseline <= 0 || current >= baseline || -delta < rule.minimumThreshold {
		return "", false
	}
	if deltaPct <= -rule.criticalThreshold {
		return modelSignalSeverityCritical, true
	}
	if deltaPct <= -rule.warningThreshold {
		return modelSignalSeverityWarning, true
	}
	return "", false
}

func (rule modelSignalDriftRule) absoluteIncreaseSeverity(current, baseline, delta float64) (string, bool) {
	if current <= baseline || current < rule.minimumThreshold {
		return "", false
	}
	if delta >= rule.criticalThreshold {
		return modelSignalSeverityCritical, true
	}
	if delta >= rule.warningThreshold {
		return modelSignalSeverityWarning, true
	}
	return "", false
}

func (field modelSignalDriftField) value(item model.ModelSignalsMetricSet) float64 {
	switch field {
	case modelSignalDriftP90Latency:
		return item.P90ModelLatencyMsPer1kOutputTokens
	case modelSignalDriftLatency:
		return item.ModelLatencyMsPer1kOutputTokens
	case modelSignalDriftP10Throughput:
		return item.P10ModelThroughputTokensPerSecond
	case modelSignalDriftOutputThroughput:
		return item.ModelThroughputOutputTokensPerSecond
	case modelSignalDriftToolFailureRate:
		return item.ToolFailureRate
	case modelSignalDriftFailurePressure:
		return item.FailurePressure
	case modelSignalDriftCacheMissRate:
		return item.CacheMissRate
	case modelSignalDriftAvgModelCalls:
		return item.AvgModelCallsPerSession
	case modelSignalDriftOutputExpansion:
		return item.OutputExpansionRate
	case modelSignalDriftReasoningOverhead:
		return item.ReasoningOverheadRate
	case modelSignalDriftDegradationRisk:
		return item.DegradationRiskScore
	default:
		return 0
	}
}

func modelSignalSampleNote(current, baseline model.ModelSignalsMetricSet) string {
	switch {
	case current.SessionCount == 0 && baseline.SessionCount == 0:
		return "missing current and baseline windows"
	case current.SessionCount == 0:
		return "missing current window"
	case baseline.SessionCount == 0:
		return "missing baseline window"
	case current.ModelCalls == 0 || baseline.ModelCalls == 0:
		return "missing model call samples"
	case current.SessionCount < 2 || baseline.SessionCount < 2:
		return "low current or baseline sample"
	default:
		return ""
	}
}

func appendModelSignalDriftMetric(drift *model.ModelSignalsDrift, metric model.ModelSignalsDriftMetric, reason string) {
	drift.Metrics = append(drift.Metrics, metric)
	drift.Reasons = appendUniqueString(drift.Reasons, reason)
	drift.Severity = worseModelSignalSeverity(drift.Severity, metric.Severity)
}

func modelSignalMetricSetLowSample(item model.ModelSignalsMetricSet) bool {
	return item.SessionCount > 0 && (item.SessionCount < 3 || item.ModelCalls < 3 || item.ModelDurationMS <= 0)
}

func modelSignalSeverityRank(severity string) int {
	switch severity {
	case modelSignalSeverityCritical:
		return 0
	case modelSignalSeverityWarning:
		return 1
	case modelSignalSeverityUnknown:
		return 2
	case modelSignalSeverityHealthy:
		return 3
	default:
		return 4
	}
}

func worseModelSignalSeverity(left, right string) string {
	if left == "" {
		return right
	}
	if right == "" {
		return left
	}
	if modelSignalSeverityRank(right) < modelSignalSeverityRank(left) {
		return right
	}
	return left
}

func addModelSignalReasonCounts(counts map[string]int, reasons []string) {
	for _, reason := range reasons {
		if strings.TrimSpace(reason) == "" {
			continue
		}
		counts[reason]++
	}
}

func topModelSignalReasons(counts map[string]int, limit int) []string {
	if len(counts) == 0 || limit <= 0 {
		return []string{}
	}
	reasons := make([]string, 0, len(counts))
	for reason := range counts {
		reasons = append(reasons, reason)
	}
	sort.Slice(reasons, func(i, j int) bool {
		if counts[reasons[i]] != counts[reasons[j]] {
			return counts[reasons[i]] > counts[reasons[j]]
		}
		return reasons[i] < reasons[j]
	})
	if len(reasons) > limit {
		reasons = reasons[:limit]
	}
	return reasons
}

func firstModelSignalReason(reasons []string) string {
	if len(reasons) == 0 {
		return ""
	}
	return reasons[0]
}

func appendUniqueString(values []string, value string) []string {
	for _, candidate := range values {
		if candidate == value {
			return values
		}
	}
	return append(values, value)
}
